mlir/lib/Dialect/Rock/Transforms/AttentionToGridwise.cpp

```diff
@@ -0,0 +1,179 @@
+
+  if (!adaptor.getQueries().getType().isa<MemRefType>())
+    return op.emitOpError("Cannot lower unbufferized gemm to gridwise");
+
+  bool isAccel = rock::isAccel(op.getFeatures());
+  if (!isAccel) {
+    return op.emitError("Currently, attention op is only supported on GPUs "
+                        "with matrix accelerator extentions");
+  }
+  Attribute params = getTuningParams(rw, op);
+
+  Value queries = adaptor.getQueries();
+  Value keys = adaptor.getKeys();
+  Value values = adaptor.getValues();
+  Value out = adaptor.getOut();
+  // Note: the gridwise ops take K x M and K x N, so A must be transposed if
+  // it's in the natural M x K form
+  queries = normalizeMatrix(queries, rw, loc, !op.getQTransposed(), "gemm0K",
+                            "gemm0M");
+  keys =
+      normalizeMatrix(keys, rw, loc, op.getKTransposed(), "gemm0K", "gemm0N");
+  values = normalizeMatrix(values, rw, loc, !op.getVTransposed(), "gemm1K",
+                           "gemm1N");
+  out = normalizeMatrix(out, rw, loc, op.getOTransposed(), "gemm1M", "gemm1N");
+
+  // Note, matrix dimension correctness is handled in the verifier
+  ArrayRef<int64_t> queriesShape =
+      queries.getType().cast<MemRefType>().getShape();
+  ArrayRef<int64_t> keysShape = keys.getType().cast<MemRefType>().getShape();
+  ArrayRef<int64_t> valuesShape =
+      values.getType().cast<MemRefType>().getShape();
+  GemmSize gemm0Size(/*g=*/queriesShape[0], /*m=*/queriesShape[2],
+                     /*k=*/queriesShape[1],
+                     /*n=*/keysShape[2]);
+  GemmSize gemm0ExtraPad =
+      requiredPadding(params, gemm0Size).value_or(GemmSize{0, 0, 0, 0});
+  GemmSize gemm1Size(/*g=*/queriesShape[0], /*m=*/queriesShape[2],
+                     /*k=*/valuesShape[1],
+                     /*n=*/valuesShape[2]);
+  GemmSize gemm1ExtraPad =
+      requiredPadding(params, gemm1Size).value_or(GemmSize{0, 0, 0, 0});
+
+  queries = padMatrix(queries, rw, loc, "gemm0K", gemm0ExtraPad.k, "gemm0M",
+                      gemm0ExtraPad.m);
+  keys = padMatrix(keys, rw, loc, "gemm0K", gemm0ExtraPad.k, "gemm0N",
+                   gemm0ExtraPad.n);
+  values = padMatrix(values, rw, loc, "gemm1K", gemm1ExtraPad.k, "gemm1N",
+                     gemm1ExtraPad.n);
+  out = padMatrix(out, rw, loc, "gemm1M", gemm1ExtraPad.m, "gemm1N",
+                  gemm1ExtraPad.n);
+
+  auto accelParams = params.cast<RockAccelTuningParamAttrInterface>();
+  int64_t waveSize = rock::lookupArchInfo(op.getArchAttr()).waveSize;
+  int64_t blockSize = waveSize * accelParams.getNPerBlock() *
+                      accelParams.getMPerBlock() /
+                      (accelParams.getMPerWave() * accelParams.getNPerWave());
+  IntegerAttr blockSizeAttr = rw.getI32IntegerAttr(blockSize);
+  int64_t gridSize = (gemm0Size.m / accelParams.getMPerBlock()) * gemm0Size.g;
+  IntegerAttr gridSizeAttr = rw.getI32IntegerAttr(gridSize);
+
+  func::FuncOp func = op->getParentOfType<func::FuncOp>();
```

Comment:
Yeah, this bit definitely goes in with the other tuning parameter affixation logic